basics/uri.adoc

```diff
@@ -330,99 +565,424 @@ Remote UUris mean that the uThing is running on a different device and we need t
+*MUST* contain the length of the `UAUTHORITY_NAME` field in bytes.
 
+|UAUTHORITY_NAME
+a|The URI's authority. Occupies bits 72 through ( 72 + 8*NAME_LEN - 1 ). +
+*MUST* contain the UTF8 encoding of the UUri's `authority.name` property.
 
-.Local Uri
-[width="80%",cols="30%,70%",options="header"]
 |===
-|Format | Value
 
-|Long
-|`/core.usubscription/3/SubscriptionChange#Update`
+NOTE: The authority is not padded to the next 32 bit word.
+--
 
-|Short
-| `/0/3/32768`
+=== Examples
 
-|Micro
-|`[1, 0, -128, 0, 0, 0, 3, 0]`
+The examples below can be used as test vectors for implementing client libraries.
 
+[%autowidth]
 |===
+|authority |entity |version |resource |Short form URI
+
+| none
+| `0x0000`
+| `0x03`
+| `0x8000`
+l|[
+  0x01, // version
+  0x00, // type
+  0x80, 0x00, // UResource ID
+  0x00, 0x00, // UEntity ID
+  0x03, // UEntity major version
+  0x00, // unused
+]
+
+l|`1G1YZ23J9P5800001`
+| `0x01ff`
+| `0x02`
+| `0x8010`
+l|[
+  0x01, // version
+  0x03, // type
+  0x80, 0x10, // UResource ID
+  0x01, 0xff, // UEntity ID
+  0x02, // UEntity major version
+  0x00, // unused
+  0x11, // length of authority
+  0x31, 0x47, 0x31, 0x59, 0x5A, 0x32, // authority
+  0x33, 0x4A, 0x39, 0x50, 0x35, 0x38, // authority
+  0x30, 0x30, 0x30, 0x30, 0x31        // authority
+]
 
+|===
 
-=== Remote Uris
 
+== Serialization
 
-.Remote IPv4 Uri
-[width="80%",cols="30%,70%",options="header"]
-|===
-|Format | Value
+Each link:../languages.adoc[uProtocol Language Library] *MUST* provide means to serialize UUris to the formats defined in the previous sections and vice versa. A concrete implementation should follow common practices for the particular programming language. For example, a Java library might implement a `UriSerializer` class providing corresponding _static_ methods.
 
-|Long
-|`//vcu.veh.gm.com/core.usubscription/3/SubscriptionChange#Update`
+[source,java]
+----
+public final class UriSerializer {
+  /**
+   * @returns The URI parsed from the given string representation.
+   * @throws UuriSerializationException if the given string is not a long form
+   *                     UUri representation. The exception may contain details
+   *                     regarding the violated constraint(s).
+   */
+  public static UUri deserializeLongFormUri(String: uri) throws UuriSerializationException {
+    ...
+  }
+  /**
+   * @returns The given URI's long form string representation.
+   * @throws UuriSerializationException if the URI cannot be serialized using the
+   *                     long form. The exception may contain details regarding
+   *                     the violated constraint(s).
+   */
+  public static String serializeLongFormUri(UUri: uri) throws UuriSerializationException {
+    ...
+  }
+}
+----
 
-|Short
-| `//192.168.1.100/0/3/32768`
+Alternatively, the `UUri` class might provide corresponding methods.
 
-|Micro
-|`[1, 1, -128, 0, 0, 0, 3, 0, -64, -88, 1, 100]`
+[source,java]
+----
+public class UUri {
```

Comment:
UUri is defined in proto and the classes are generated to which we cannot extend. I recommend we do not show coding examples in this spec but state that serializers and deserializers and validators are required.

Reply:
ok, so simply remove the code examples?